doc/src/ch02-lang/ch02-expressions.md

```diff
@@ -0,0 +1,95 @@
+# Expressions
+
+## `push` opcodes
+
+The `push` opcodes are the only EVM opcodes that have an immediate argument. This means that instead of taking an argument off the stack, the opcode reads the bytes from code which immedaitely succeed the op.
+
+For example, `push2 258` would assemble to `0x610102`. `push2` notifies the EVM that it should use the next two bytes `0x01` and `0x02` as input for the op. That value is left-padded to 32 bytes and placed on the stack.
+
+While an assembled `push` must have a concrete value, it is often useful when developing a program to have the ability to manipulate the operand at compile time. For this reason, `push` opcodes take a single **expression** as their operand.
+
+```rust
+# extern crate etk_asm;
+# let src = r#"
+push1 1+(2*3)/4
+# "#;
+# let mut output = Vec::new();
+# let mut ingest = etk_asm::ingest::Ingest::new(&mut output);
+# ingest.ingest(file!(), src).unwrap();
+# assert_eq!(output, &[0x60, 0x02]);
+```
+
+## Definition
+
+An **expression** is a standard infix mathematical expression that is evaluated during assembly. It's computed value *must* fit within the preceeding `push`'s size allowance. 
+
+### Terms
+
+Many different types of values are allowed as a term in an expression:
+
+#### Integer Literals
+
+Integer literals are described by the following Pest grammar:
+
+```ignore
+number   = _{ binary | octal | hex | decimal | negative }
+decimal  = @{ ASCII_DIGIT+ }
+negative = @{ "-" ~ ASCII_DIGIT+ }
+binary   = @{ "0b" ~ ASCII_BIN_DIGIT+ }
+hex      = @{ "0x" ~ ASCII_HEX_DIGIT ~ ASCII_HEX_DIGIT+ }
+octal    = @{ "0o" ~ ASCII_OCT_DIGIT+ }
+```
+
+There is no limit for the length of integer literals.
+
+#### Labels
+
+A [label](ch03-labels.md) may be used as a term in an expression.
+
+```rust
+# extern crate etk_asm;
+# let src = r#"
+start:
+    push1 start + 1
+# "#;
+# let mut output = Vec::new();
+# let mut ingest = etk_asm::ingest::Ingest::new(&mut output);
+# ingest.ingest(file!(), src).unwrap();
+# assert_eq!(output, &[0x60, 0x01]);
+```
+
+#### Macros
+
+[Expression macros](ch04-macros/ch02-expression.md) may be used as a term in an expression.
+
+```rust
+# extern crate etk_asm;
+# let src = r#"
+push4 selector("transfer(uint256,uint256)")
+# "#;
+# let mut output = Vec::new();
+# let mut ingest = etk_asm::ingest::Ingest::new(&mut output);
+# ingest.ingest(file!(), src).unwrap();
+# assert_eq!(output, &[0x63, 12, 247, 158, 10]);
+```
+
+### Operators
+
+#### Binary
+
+Expressions support the following binary operators:
```

Comment:
Might be important to note that we support signed and unsigned, and their encodings?

Reply:
Right - I noted in the intersection in the definition, but probably valuable to state explicitly.